Write stats to disk and `dolt_statistics` table

max-hoffman opened the pull request:
New flatbuffer type for statistics, which contains an address reference to a prolly.Map with a hardcoded statistics schema. Statistics for a table are rewritten every ANALYZE TABLE call.

Because the statistics schema is client-dependent, backwards compatibility will be fairly brittle right now. We could use a version identifier to invalidate a set of statistics if the client code creating a sql engine has a different version than the statistics read from disk.

nicktobey commented on go/libraries/doltcore/ref/stats_ref.go:
```diff
@@ -0,0 +1,44 @@
+
+var _ DoltRef = StatsRef{}
+
+// NewStatsRef creates a reference to a stashes list. There cannot be more than one statsRef.
```
At first I thought that "stash" was a typo of "stats" but it's consistent enough that now I'm not sure. Can you clarify why we're calling these "stashes"?

max-hoffman replied:
sorry bad copy paste, will fix

max-hoffman commented on go/libraries/doltcore/doltdb/doltdb.go:
```diff
@@ -136,8 +137,8 @@ func (ddb *DoltDB) AccessMode() chunks.ExclusiveAccessMode {
 }
 
 // CommitRoot executes a chunkStore commit, atomically swapping the root hash of the database manifest
-func (ddb *DoltDB) CommitRoot(ctx context.Context, last, current hash.Hash) (bool, error) {
```
yeah, it being reverse compared to the other commit interfaces it depends on is kind of confusing imo

max-hoffman replied:
Unless I'm missing something the change is all internal to this function call though, and makes it consistent with the other interfaces

nicktobey commented on go/serial/stat.fbs:
```diff
@@ -0,0 +1,28 @@
+
+namespace serial;
+
+// StoreRoot is the top-level chunk in the store,
```
Replace this docstring?

nicktobey commented on go/serial/stat.fbs:
```diff
@@ -0,0 +1,28 @@
+
+// StoreRoot is the top-level chunk in the store,
+// containing references to all named Refs.
+table Statistic {
```
Is the plan to add more fields to this over time? Otherwise I'm not sure why the root doesn't just references the table address directly instead of going through this.

Unless I'm misunderstanding the purpose of this message type.

max-hoffman replied:
Aaron and I talked about how it might be better to have a dedicated message type for this, otherwise the helper functions that limit the acceptable root types would have to include the generic prolly node message type.

max-hoffman commented:
I realized last night that I need to tweak the update to not erase stats for other tables. Will add a fix where I delete all rows from the map for the given qualifier before re-adding.

nicktobey commented on go/libraries/doltcore/sqle/stats/stats_provider.go:
```diff
@@ -164,32 +172,83 @@ type indexMeta struct {
 
 var _ sql.StatsProvider = (*Provider)(nil)
 
+// Init scans the statistics tables, populating the |stats| attribute.
```
Init -> Load

nicktobey commented on go/libraries/doltcore/sqle/stats/stats_provider.go:
```diff
@@ -164,32 +172,83 @@ type indexMeta struct {
+			} else if err != nil {
+				return err
+			}
+			p.dbStats[dbName] = stats
```
It's not safe to modify maps concurrently: https://go.dev/blog/maps#TOC_6.

The official suggestion here is to use a mutex, but I suspect it would also be safe to insert nil values for each db name before creating any of the goroutines, and then reassigning those values in the goroutines. That should guarentee that the map doesn't get resized while another routine is writing to it.

nicktobey replied:
I think I misread; if p.dbStats[dbName] already exists for each dbName this is probably safe.

nicktobey commented on go/libraries/doltcore/sqle/stats/write.go:
```diff
@@ -0,0 +1,105 @@
+
+func newStatsTable(ctx *sql.Context, ns tree.NodeStore, vrw stypes.ValueReadWriter) (*doltdb.Table, error) {
+	//  sqlutil.ToDoltSchema(ctx, root, tableName, sch, headRoot, collation)
+	return doltdb.CreateEmptyTable(ctx, ns, vrw, schema.StatsTableDoltSchema)
```
Remove this comment?

nicktobey commented on go/libraries/doltcore/sqle/stats/write.go:
```diff
@@ -0,0 +1,105 @@
+	return doltdb.CreateEmptyTable(ctx, ns, vrw, schema.StatsTableDoltSchema)
+}
+
+// TODO: write a new set of statistics to the stats_provider
```
Clean up this docstring?

nicktobey reviewed (commented):
Looks good!

I think the only thing missing is a test that we can write stats out to disk and read them back in again.

max-hoffman commented on go/libraries/doltcore/schema/statistic.go:
```diff
@@ -0,0 +1,107 @@
+
+func StatsTableDoltSchemaGen() Schema {
+	colColl := NewColCollection(
+		NewColumn(statsQualifier, qualifierTag, stypes.StringKind, true, NotNullConstraint{}),
```
Not super opinionated, seems like what other databases do, flexible for describing a physical path to the datastore for a set of buckets. It's maybe permissive if we want something other than db/table/index as the data source.

nicktobey replied:
Sounds good.

max-hoffman commented:
I did some heavy organizational changes to the dolt stats PR, added `dolt_statistics` table, split out qualifier pieces into their own columns, added bunch of tests. Going to swing back to the GMS PR first to review, once that one's merged I'll do the Dolt side

nicktobey commented on go/libraries/doltcore/sqle/dtables/statistics_table.go:
```diff
@@ -0,0 +1,268 @@
+	"time"
+)
+
+// StatisticsTable is a sql.Table implementation that implements a system table which shows the dolt commit log
```
"shows the dolt commit log" -> Update this?

nicktobey commented on go/libraries/doltcore/sqle/enginetest/stats_queries.go:
```diff
@@ -299,20 +303,83 @@ var DoltHistogramTests = []queries.ScriptTest{
+			{
+				Query: "select `database`, `table`, `index`, commit_hash, columns, types from dolt_statistics",
+				Expected: []sql.Row{
+					{"mydb", "xy", "PRIMARY", hash.Parse("f6la1u3ku5pucfctgrca2afq9vlr4nrs"), []string{"x"}, []string{"bigint"}},
```
I feel like hard coding the hash into the test makes it brittle and unreadable. The test should construct the value we expect it to point to and take the hash of that value.

nicktobey commented on go/libraries/doltcore/sqle/statistics_provider.go:
```diff
@@ -0,0 +1 @@
+package sqle
```
Is this file necessary?

nicktobey reviewed (approved):
Moving the stats logic into a custom `sql.Table` was a good idea.

Tests look good.

Approved.

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997676</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>6ca6950</TD>
<TD>not ok</TD>
<TD>138</TD>
</TR>
<TR><TD>6ca6950</TD>
<TD>ok</TD>
<TD>5937319</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>6ca6950</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997676</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997676</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>b356903</TD>
<TD>not ok</TD>
<TD>138</TD>
</TR>
<TR><TD>b356903</TD>
<TD>ok</TD>
<TD>5937319</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>b356903</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997676</TD>
</TR>
</table>

coffeegoddd commented:
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997676</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>3e4bc44</TD>
<TD>not ok</TD>
<TD>138</TD>
</TR>
<TR><TD>3e4bc44</TD>
<TD>ok</TD>
<TD>5937319</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>3e4bc44</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997676</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997709</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>96ee545</TD>
<TD>not ok</TD>
<TD>136</TD>
</TR>
<TR><TD>96ee545</TD>
<TD>ok</TD>
<TD>5937321</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>96ee545</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997709</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997709</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>61bedc9</TD>
<TD>not ok</TD>
<TD>136</TD>
</TR>
<TR><TD>61bedc9</TD>
<TD>ok</TD>
<TD>5937321</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>61bedc9</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997709</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997709</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>67dd08e</TD>
<TD>not ok</TD>
<TD>136</TD>
</TR>
<TR><TD>67dd08e</TD>
<TD>ok</TD>
<TD>5937321</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>67dd08e</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997709</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997709</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>5b0936c</TD>
<TD>not ok</TD>
<TD>136</TD>
</TR>
<TR><TD>5b0936c</TD>
<TD>ok</TD>
<TD>5937321</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>5b0936c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997709</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.998097</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d6e28ed</TD>
<TD>not ok</TD>
<TD>113</TD>
</TR>
<TR><TD>d6e28ed</TD>
<TD>ok</TD>
<TD>5937344</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d6e28ed</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998097</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.998097</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>cd83ac6</TD>
<TD>not ok</TD>
<TD>113</TD>
</TR>
<TR><TD>cd83ac6</TD>
<TD>ok</TD>
<TD>5937344</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>cd83ac6</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998097</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.998097</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>305796c</TD>
<TD>not ok</TD>
<TD>113</TD>
</TR>
<TR><TD>305796c</TD>
<TD>ok</TD>
<TD>5937344</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>305796c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998097</TD>
</TR>
</table>
